### ISSUE TITLE
Problem using Read/WritePasswordJob with QtConcurrentRun

### ISSUE THREAD

**Issue by ThanasisPap** [frankosterfeld/qtkeychain#145]
I tried to integrate some functionality of this library with QtConcurrent. Unfortunately the application crashes with Read Access Violation error. 

**SAMPLE CODE:**

```
KeyHandler keyHandler;
keyHandler.setPassword("12345");

QFuture<std::string> f = QtConcurrent::run(&keyHandler, &KeyHandler::getKey, KEY_NAMES::KEY_1);
while(!f.isFinished())
     QThread::msleep(5);

qDebug()<<"Finished!";
```
```
void KeyHandler::setPassword(const std::string& pass)
{
    std::unique_ptr<QKeychain::WritePasswordJob> write = std::make_unique<QKeychain::WritePasswordJob>("");
    write->setSettings(&sett);
    write->setKey(KEY_NAMES::PASS);
    write->setTextData(QString::fromStdString(pass));

    QEventLoop loop;
    write->connect(write.get(), &QKeychain::WritePasswordJob::finished, &loop, &QEventLoop::quit);

    write->start();
    loop.exec();

    if(write->error())
        qDebug()<<"Error writing key1. Error: "<<write->errorString();
}
```
```
std::string KeyHandler::getKey(const QString& name)
{
    std::unique_ptr<QKeychain::ReadPasswordJob> readPass = std::make_unique<QKeychain::ReadPasswordJob>("");
    readPass->setAutoDelete(true);
    readPass->setKey(name);
    readPass->setSettings(&sett);

    QEventLoop loop;
    readPass->connect(readPass.get(), &QKeychain::ReadPasswordJob::finished, &loop, &QEventLoop::quit);

    readPass->start();
    loop.exec();

    if(readPass->error())
    {
        qDebug()<<"KeyHandler~Error reading decrypt Pass. Error: " + readPass->errorString();
        return "";
    }
    else
    {
        std::string data = readPass->textData().toStdString();
        return data;
    }
}
```
The above code reaches "Finished!" and then crashes with Read Access Violation error.

**Comment by d1vanov** [comment]
I believe the problem is caused by the fact that by default QtKeychain's jobs delete themselves but you wrap them into `std::unique_ptr` which **also** attempts to delete them when they go out of scope which is in the end of execution. You should either not use `std::unique_ptr` for QtKeychain's jobs or you should call `setAutoDelete(false)` on both write and read jobs before starting them.

**Comment by ThanasisPap** [comment]
I have already tried both of your proposed solutions. I've created them on the stack and set AutoDelete to false. Despite these changes the application still crashes. Here is a sample code to reproduce it:

```
#include <QCoreApplication>

#include "qt5keychain/qkeychain_export.h"
#include "qt5keychain/keychain.h"

#include <QtConcurrent/qtconcurrentrun.h>
#include <QFuture>
#include <QEventLoop>

#include <iostream>
#include <QDebug>
#include <QSettings>
#include <string>

namespace KEY_NAMES {
    static const QString KEY_1 = "key1";
    static const QString PASS = "pass";
}

static const QString APP_NAME = "AppName";
static const QString ORG_NAME = "OrgName";

static QSettings sett{ORG_NAME,APP_NAME};

void setPassword(const std::string& pass)
{
    QKeychain::WritePasswordJob write("");
    write.setSettings(&sett);
   // write.setAutoDelete(false);
    write.setKey(KEY_NAMES::PASS);
    write.setTextData(QString::fromStdString(pass));

    QEventLoop loop;
    write.connect(&write, &QKeychain::WritePasswordJob::finished, &loop, &QEventLoop::quit);

    write.start();
    loop.exec();

    if(write.error())
        qDebug()<<"Error writing key1. Error: "<<write.errorString();
}

std::string getKey(const QString& name)
{
    QKeychain::ReadPasswordJob readPass("");
    readPass.setSettings(&sett);
   // readPass.setAutoDelete(false);
    readPass.setKey(name);

    QEventLoop loop;
    readPass.connect(&readPass, &QKeychain::ReadPasswordJob::finished, &loop, &QEventLoop::quit);

    readPass.start();
    loop.exec();

    if(readPass.error())
    {
        qDebug()<<"KeyHandler~Error reading decrypt Pass. Error: " + readPass.errorString();
        return "";
    }
    else
    {
        std::string data = readPass.textData().toStdString();
        return data;
    }
}

int main(int argc, char *argv[])
{
    QCoreApplication a(argc, argv);

    setPassword("12345");

    QFuture<std::string> f = QtConcurrent::run(&getKey, KEY_NAMES::KEY_1);
    while(!f.isFinished())
        QThread::msleep(5);

    qDebug()<<"Finished!";

    return a.exec();
}
```

**Comment by frankosterfeld** [comment]
On which platform are you trying? If it reaches Finished, it doesn't look like the job execution or destruction itself is the problem here...

**Comment by frankosterfeld** [comment]
Also, any useful backtrace when running it in a debugger?

**Comment by ThanasisPap** [comment]
@frankosterfeld I am building using MSVC 2017 64bit, Qt5.11.1 on a Windows 10 PC. As I mentioned before, the app reach "Finished!" all the times.  I just figure out that If you modify main's :
`return a.exec();`
to 
`return 0;`
surprisingly it doesn't crash.

The thing is that I have a different implementation where I use QtConcurrent::run and the things are pretty strange. When I make multiple calls on QKeyChain's objects the executing thread never finishes (the application stalls). 

With the example above I was just trying to reproduce the error but it seems that I didn't perfectly match it. I 'll try to make a more clear example to highlight the issue. 

Thank you.

**Comment by ThanasisPap** [comment]
Just replace the final sector of the source code that I posted before to reproduce it:

```
void doStuff()
{
    qDebug()<<"Get Key 1";
    getKey(KEY_NAMES::KEY_1);
    qDebug()<<"Get Key 1 Successfull";
    qDebug()<<"Get key 2";
    getKey(KEY_NAMES::KEY_1);
    qDebug()<<"Get Key 2 Successfull";
}

int main(int argc, char *argv[])
{
    QCoreApplication a(argc, argv);

    setPassword("12345");

    qDebug()<<"Start Qtconcurrent";

    QFuture<void> f = QtConcurrent::run(&doStuff);
    while(!f.isFinished())
        QThread::msleep(5);

    qDebug()<<"Finished!";

    return 0;
}

```

The program stalls at second call of getKey (whatever the QKeyChain call is i.e. if you replace getKey with sePassword you will have the exact same behavior). The QEventLoop never quits. 

Note 1: If you remove setPassword("12345") in main the application runs successfully.

**Comment by fbaune** [comment]
Hi,

I had the same issue using QtKeychain jobs in a different thread.

I was declaring and using a ReadJob in a function (the ReadJob objet was implicitly destroyed at the end of the function).
I tried to change AutoDelete to true or false but nothing changed.

I finally fixed the problem by creating ReadJob object as a pointer (with a new statement) without implement delete (I checked, object is destroyed by the system).

Note: I tried to add a deleteLater call on ReadJob object but the problem reappearded.

My guess : a deleteLater (or equivalent behaviour) is sent to the main eventloop by QtKeychain library. In your function, I think your Job is destroyed at the end of the function, but ! The deleteLater event is still in the eventloop and when it resolve.. ReadError.

I hope it'll help,
Regards

**Comment by ThanasisPap** [comment]
> Hi,
> 
> I had the same issue using QtKeychain jobs in a different thread.
> 
> I was declaring and using a ReadJob in a function (the ReadJob objet was implicitly destroyed at the end of the function). I tried to change AutoDelete to true or false but nothing changed.
> 
> I finally fixed the problem by creating ReadJob object as a pointer (with a new statement) without implement delete (I checked, object is destroyed by the system).
> 
> Note: I tried to add a deleteLater call on ReadJob object but the problem reappearded.
> 
> My guess : a deleteLater (or equivalent behaviour) is sent to the main eventloop by QtKeychain library. In your function, I think your Job is destroyed at the end of the function, but ! The deleteLater event is still in the eventloop and when it resolve.. ReadError.
> 
> I hope it'll help, Regards

Your mentions are correct. I have had to figure it out since the main event loop was stalling.
However I switched to direct usage of dpapi since I deploy only on Windows and I've had relatively simple tasks to perform. No concurrency related issues till then.

Thank you.
